Quilt (& intermediary) support

anonymous123-code opened the pull request:
- Intermediary and Quilt support
- Does not add stuff required for dependency stuff (And thus proper launcher support, although its usable with manual component editing)
- No caching (sha and content-length require two requests per library)
- It uses the corresponding last-modified header, although those arent the most reliable (especially quilt's has a cutoff at 2023-04-30)

anonymous123-code committed:
Quilt?

anonymous123-code committed:
Fixes not getting the correct size
setting the wrong url (was maven root)
Also the first special casing for broken stuff

anonymous123-code committed:
Actually add the loader to downlaods & classpath

anonymous123-code committed:
Use meta instead of maven meta
(no xml parsing needed anymore)

anonymous123-code committed:
Fix: Intermediary used Utc::now as release date

anonymous123-code committed:
Update User-Agent header

kb-1000 commented:
Okay so, the thing here is that I'm not sure if I'd want to run this code twice per hour without caching. How's the performance on this?

anonymous123-code commented:
It likely isn't good enough tbh. The code takes like 10 minutes (even when rerunning). I will change this to use caching when I have more time.

anonymous123-code committed:
Rework the quilt meta generation:
- Lots of async (It felt like caching isnt even needed anymore)
- Lots of caching (Best case only the request for update checking is needed)

anonymous123-code committed:
Rework the intermediary meta generation:
- Lots of async
- Lots of caching (Best case only the request for update checking is needed)
- cargo fmt

anonymous123-code commented:
I now reworked everything to be concurrent and implemented caching. For downloads, the hashes are stored per version, not per download version. So there is room for more caching, although I doubt it is necessary. https://helixlauncher-meta-test.pages.dev/

anonymous123-code commented:
Note that because Im using `buffer_unordered`, the download Vec order is inconsistent, although this is cached. If this is a problem, it can be changed to `buffered`.
